rclcpp/include/rclcpp/parameter_client.hpp

```diff
@@ -73,28 +74,32 @@ class AsyncParametersClient
     std::function<void(
       std::shared_future<std::vector<rclcpp::parameter::ParameterVariant>>)> callback = nullptr)
   {
-    std::promise<std::vector<rclcpp::parameter::ParameterVariant>> promise_result;
-    auto future_result = promise_result.get_future().share();
+    auto promise_result =
+      std::make_shared<std::promise<std::vector<rclcpp::parameter::ParameterVariant>>>();
+    get_parameters_promises_.insert(promise_result);
+    auto future_result = promise_result->get_future().share();
 
     auto request = std::make_shared<rcl_interfaces::srv::GetParameters::Request>();
     request->names = names;
 
     get_parameters_client_->async_send_request(
       request,
-      [&names, &promise_result, &future_result, &callback](
+      [names, promise_result, &future_result, &callback, this](
```

Comment:
What is `this` used for in the lambda function? Should the lambda body use `this` explicitly to make it more visible? Or instead should other variables being specified in the capture list?